feat: Add tutorial for Permissionless Multisigner paymaster (#37)
zkSync-Community-Hub/community-code: content/tutorials/permissionless-paymaster/10.index.md

---
title: Integrate permissionless multi-signer paymaster into your Dapp
description: Permissionless paymaster allows multiple Dapps to seamlessly sponsor gas for their users using signature verification.
---

This tutorial shows you how to integrate this paymaster in your Dapp.

You will:

- Create a signer address that will sign paymaster data for gas sponsorship.
- Deposit gas funds and add the signer address.
- Sign EIP-712 standard signature on the paymaster data by the signer.
- Send the transaction with paymaster data & signature.

## Prerequisites

> - Make sure you have a basic front-end integration ready, you can create a standard frontend [using zksync-cli](https://docs.zksync.io/build/zksync-cli/creating-projects).

Reviewer: This might be too much to do, but it would be ideal if you can say exactly what starter template to use and provide the instructions from there that are more specific to the template project

> - Make sure you have a basic front-end integration ready, you can create a standard frontend [using zksync-cli](https://github.com/matter-labs/era-compiler-solidity/tree/main#system-requirements).

Reviewer 2: Is this expected link for the cli?

Author: It's a mistake. Have updated with the correct link, ty!

- A [Node.js](https://nodejs.org/en/download) installation running at minimum Node.js version 18.
- You are already familiar with paymaster integration on ZKsync Era.
  If not, please refer to the first section of the [paymaster introduction](https://docs.zksync.io/build/quick-start/paymasters-introduction).
- Some background knowledge on the concepts covered by the tutorial would be helpful too. Have a look at the following
  docs:
  - [EIP-712 standard for typed message signing](https://eips.ethereum.org/EIPS/eip-712).

>   - [Transaction lifecycle](https://docs.zksync.io/zk-stack/concepts/transaction-lifecycle#eip-712-0x71) on zkSyn Era.

Reviewer 3: ```suggestion
  - [Transaction lifecycle](https://docs.zksync.io/zk-stack/concepts/transaction-lifecycle#eip-712-0x71) on ZKsync Era.
```

  - [Gas estimation for transactions](https://docs.zksync.io/build/developer-reference/fee-model#gas-estimation-for-transactions) guide.
  - [Introduction to system contracts especially NonceHolder](https://docs.zksync.io/build/developer-reference/era-contracts/system-contracts#nonceholder).
- You know how to get your [private key from your MetaMask wallet](https://support.metamask.io/hc/en-us/articles/360015289632-How-to-export-an-account-s-private-key).

## Project

> The similar integration code is [available on GitHub here](https://github.com/ondefy/permissionless-multisigner-paymaster/blob/main/deploy/interact.ts).

Reviewer: ```suggestion
A similar version of the integration code is [available on GitHub here](https://github.com/ondefy/permissionless-multisigner-paymaster/blob/main/deploy/interact.ts).
```

::callout{icon="i-heroicons-exclamation-circle"}

> This tutorial does not contain the entire frontend code require for the Dapp.

Reviewer: ```suggestion
This tutorial does not contain the entire frontend code required for the Dapp.
```

> Instead it covers all the important part required by the Dapp to integrate this paymaster easily.

Reviewer: ```suggestion
Instead it covers just the important parts required by the Dapp to integrate this paymaster easily.
```

::

## Overview
The permissionless multi-signer paymaster serves as a public good, enabling ZKsync Dapps to seamlessly sponsor gas for their users through signature verification.

> Dapps can begin utilizing this paymaster by simply depositing funds and adding a signer address. Thus removing the need to deploy paymaster at all.

Reviewer: ```suggestion
Dapps can begin utilizing this paymaster by simply depositing funds and adding a signer address. Thus removing the need to deploy a paymaster at all.
```

There are 2 primary actors involved:

**Manager**: Fully managed by the Dapp, responsible for depositing/withdrawing gas funds, and adding or removing signer addresses.

**Signers**: Managed by the Dapp or a trusted third party like Zyfi API. A signer’s signature is required to access gas funds by the Dapp’s user.

## Multi-signer

> This paymaster allows manager to set multiple signers through which users can have access to the gas funds. Hence, one-to-many relantionship.

Reviewer: ```suggestion
This paymaster allows the manager to set multiple signers through which users can have access to the gas funds. Hence, it is a one-to-many relationship.
```

![manager-signer-relation-diagram](/images/permissionless-paymaster/manager-signer.jpg)

## Integration

> Below diagram provides the flow of the integration:

Reviewer: ```suggestion
Below the diagram provides the flow of the integration:
```

1. Dapp decides on custom logic for each user. Let's assume that Dapp decides to sponsor gas for every approve transaction.

2. Dapp calls backend server or Zyfi API with relevant data to get the signer's signature.
    - It is recommended that signer's signing part is done on a secure backend server of the Dapp.

3. The signer's key signs this paymaster data and returns the signature and signer address to the Dapp's frontend.

4. Paymaster address and required data with signature is added to the transaction blob in the frontend.

5. User gets transaction signature request pop-up on their wallet. User only signs the transaction and transaction is sent on-chain.

6. The paymaster validates the signature, identifies the manager related to the signer,
deducts gas fees from the manager's balance, and pays for the user's transaction

![flow](/images/permissionless-paymaster/flowDiagram.jpg)

For this tutorial, we will use paymaster deployed on ZKsync sepolia testnet : [0xc1B0E2edC4cCaB51A764D7Dd8121CBf58C4D9E40](https://sepolia.explorer.zksync.io/address/0xc1B0E2edC4cCaB51A764D7Dd8121CBf58C4D9E40#transactions)

## 1. Create a signer

> Paymaster will verify signature based on this signer address.

Reviewer: ```suggestion
The paymaster will verify signature based on this signer address.
```

> Private key of this signer address should be stored securely by the Dapp.

Reviewer: ```suggestion
The private key of this signer address should be stored securely by the Dapp.
```

> - Easy way to create one:

Reviewer: ```suggestion
- Here is an easy way to create one:
```

```javascript
    import { Wallet } from "zksync-ethers";
    const signer = Wallet.createRandom();
```

## 2. Deposit gas funds and add the signer address
Navigate to ZKsync Sepolia Testnet Explorer: [0xc1B0E2edC4cCaB51A764D7Dd8121CBf58C4D9E40](https://sepolia.explorer.zksync.io/address/0xc1B0E2edC4cCaB51A764D7Dd8121CBf58C4D9E40#contract).

> Call `depositAndAddSigner()` function with 0.01 ether and the signer address.

Author: Makes sense, added more clarity here.

![deposit and add a signer](/images/permissionless-paymaster/depositAndAddSigner.png)

::callout{icon="i-heroicons-exclamation-circle"}

> The depositor address is considered "manager". A manager can deposit/withdraw gas funds & add/remove signers at any given time.

Reviewer: ```suggestion
The depositor address is considered a "manager". A manager can deposit/withdraw gas funds and add/remove signers at any given time.
```

*A manager can be a signer address too. (not recommended)*
::

## 3. Create function to sign EIP-712 typed paymaster data

> This paymaster verifies signature signed by the signer address on the below data.

Reviewer: ```suggestion
This paymaster verifies the signature signed by the signer address on the below data.
```

On successful validation, it allows sponsorship for the user using manager's deposited gas funds.

```solidity
        bytes32 messageHash = keccak256(
            abi.encode(
                SIGNATURE_TYPEHASH,
                _from,
                _to,
                _expirationTime,
                _maxNonce,
                _maxFeePerGas,
                _gasLimit
            )
        );
```

> - Create `getSignature()` function in your backend(recommended)

Reviewer: ```suggestion
- Create a `getSignature()` function in your backend(recommended)
```

```javascript
// Example code

> import {BigNumber, Contract, Wallet} from "zksync-ethers";

Reviewer: Can you add the package dependencies somewhere? For example, `zksync-ethers` for this. Also `BigNumber` is not found when I try to use this code

Author: Noted, added package dependencies and updated BigNumber error.

export async function getSignature(
  from: string, to: string, expirationTime: BigNumber, maxNonce: BigNumber, maxFeePerGas: BigNumber, gasLimit: BigNumber, paymaster: Contract
){

>   const signer = new Wallet(process.env.SIGNER_PRIVATE_KEY, provider);

Reviewer: Can you clarify that the reader will need to have the provider and .env file set up for this? 

Could also add something like 

```ts
const rpcUrl = process.env.ZKSYNC_RPC_URL ?? 'https://sepolia.era.zksync.dev';
  const provider = new Provider(rpcUrl);
```

Author: Updated with provider and added more clarity on step-3

  }
// Note: MaxNonce allows the signature to be replayed.
// For eg: If currentNonce of user is 5, maxNonce is set to 10. Signature will allowed to replayed for nonce 6,7,8,9,10 on the same `to` address by the same user.
// This is to provide flexibility to Dapps to ensure signature works if users have multiple transactions running. 
// Important: Signers are recommended to set maxNonce as current nonce of the user or as close as possible to ensure safety of gas funds.
// Important : Signers should set expirationTime is close enough to ensure safety of funds.

// Signer wallet will already defined in the code
  return [(await signer._signTypedData(domain, types, values)), signer.address];
}
```

## 4. Add paymaster data and signature into transaction blob

- Before sending user transaction, you will need to add customData related to paymaster.
- Funds are deducted as per : `gasPrice`*`gasLimit`. Dapp should ensure enough funds are deposited to avoid failures.
- Note: For simplicity purposes, we use "DappContract" variable which is expected to be defined as per Dapp's need.

```javascript

> // This is example code. Direct copy/paste won't work

Reviewer: Can you explain more why this code won't work if you copy/paste it and what exactly needs to be changed to make it work?

const innerInput = ethers.utils.arrayify(
      abiCoder.encode(
["uint256","uint256","address","bytes"], 
[expirationTime, // As used in above signature 
 maxNonce, // As used in above signature
 signerAddress, // The signer address
 signature]), // Signature created in the above snippet. get from API server
    );
   // getPaymasterParams function is available in zksync-ethers
const paymasterParams = utils.getPaymasterParams(
            paymasterAddress, // Paymaster address
            {
                type: "General",
                innerInput: innerInputs
            });
// Send the transaction with paymaster data. 
// Users will get transaction signature pop-up
const tx = await DappContract.<function>([args..],{
    maxFeePerGas, // Ensure it's same as used for signature
    gasLimit, // Ensure it's same as used for signature
    customData:{
       paymasterParams, // Paymaster address + paymaster data with signature.
       gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
    },
  });

```

::callout{icon="i-heroicons-exclamation-circle"}

- The gas funds are deducted from the manager's balance related to the signer on successful verification.

> - Hence, if signer's private key is leaked, respective manager will need to replace/remove the signer immediately.

Reviewer: ```suggestion
- Hence, if a signer's private key is leaked, the respective manager will need to replace/remove the signer immediately.
```

::

> - Adding correct customData(paymasterParams : paymaster address & innerInputs), user shall get signature request pop-up for transaction.

Reviewer: ```suggestion
- Adding correct the `customData(paymasterParams : paymaster address & innerInputs)`, the user shall get a signature request pop-up for transaction.
```

> Here user can verify that gas is not being paid from their end and only signature is required.

Reviewer: ```suggestion
Here the user can verify that gas is not being paid from their end and only a signature is required.
```

    ```bash
    yarn
    yarn dev
    ```

Navigate to `http://localhost:3000/` in a browser to see the application running.

![frotend](/images/permissionless-paymaster/frontend.png)

This is a simple template that allows users to approve a certain amount of DAI tokens to a spender address on ZKsync Sepolia Testnet.

*Our goal for this tutorial is to sponsor this approval transaction using the paymaster.*

---

## 4. Create a function to sign EIP-712 typed paymaster data
This paymaster verifies the signature signed by the signer address on the below data.
On successful validation, it allows sponsorship for the user using manager's deposited gas funds.

```solidity
 bytes32 messageHash = keccak256(
 abi.encode(
 SIGNATURE_TYPEHASH,
 _from,
 _to,
 _expirationTime,
 _maxNonce,
 _maxFeePerGas,
 _gasLimit
 )
 );
```

::callout{icon="i-heroicons-exclamation-triangle"}
For the sake of this tutorial, we will be creating the signing function in the repo itself at *"src/backend/getSignatureApiCall.tsx"* path.
<br/><br/>
**It is highly recommended, that signing should be done in a closed backend server for the safety of signer's private key.**
::

> - Create a `.env` file in the root folder and add signer's private key as `NEXT_PUBLIC_SIGNER_PRIVATE_KEY`. Run `source .env` once done.

Author: I tried to support `getStaticProps` and other data fetching interfaces. The template would require quite a lot of changes. If it's okay, we can move forward since we have cautioned it in the step.

      expirationTime,
      maxNonce,
      maxFeePerGas,
      gasLimit
 );
    console.log("Signer: " + signerAddress);
    // We encode the extra data to be sent to paymaster
    // Notice how it's not required to provide from, to, maxFeePerGas, and gasLimit as per the signature above.
    // That's because paymaster will get it from the transaction struct directly to ensure it's the correct user.
    const innerInput = ethers.utils.arrayify(
      abiCoder.encode(
 ["uint256", "uint256", "address", "bytes"],
 [
          expirationTime, // As used in the above signature
          maxNonce, // As used in the above signature
          signerAddress, // The signer address
          signature,
 ]
 ) // Signature created in the above snippet. get from API server
 );
    // getPaymasterParams function is available in zksync-ethers
    const paymasterParams = utils.getPaymasterParams(
      paymasterAddress, // Paymaster address
 {
        type: "General",
        innerInput: innerInput,
 }
 );
    // Returns paymaster params, gas fee, gas limit
    return [paymasterParams, maxFeePerGas, gasLimit];
 };
```

##### **5.2** Estimate gas and call the above function

> ```javascript [src/components/WriteContract.tsx]

Reviewer: can you clarify where this is supposed to go in the file?

  // Estimate gas
    const estimateGas = await contract.estimateGas.approve(spender,amount);

>     const [paymasterParams, maxFeePerGas, gasLimit] = await preparePaymasterParam(account, estimateGas);

Reviewer: it seems like this line of code is also needed for the `account` variable: `const { account, getSigner, getProvider, network } = useEthereum();`

Author: Added more description and comments for this step. Since, we are adding in the file, those variables are already defined in the template.
